appinventor/appengine/src/com/google/appinventor/client/utils/Promise.java

```diff
@@ -0,0 +1,316 @@
+   */
+  @JsFunction
+  public interface PromiseCallback<T, R> {
+    Promise<R> resolved(T response);
+  }
+
+  /**
+   * Interface to represent a callback that will receive an error when a Promise
+   * is rejected.
+   */
+  @JsFunction
+  public interface PromiseErrorCallback {
+    @SuppressWarnings("unused")  // invoked from native code
+    Object rejected(WrappedException t);
+  }
+
+  /**
+   * Interface representing the Promise's native resolve callback.
+   *
+   * @param <T> the type of the resolved value
+   */
+  @JsFunction
+  public interface ResolveCallback<T> {
+    void apply(T arg);
+  }
+
+  /**
+   * Interface representing the Promise's native reject callback.
+   */
+  @JsFunction
+  public interface RejectCallback {
+    void apply(WrappedException arg);
+  }
+
+  /**
+   *
+   * <b>We can replaces this with Callback with GWT 2.8.2.</b>
+   *
+   * @param <T> the type of the resolved value
+   */
+  @FunctionalInterface
+  @JsFunction
+  public interface PromiseConstructorCallback<T> {
+    void call(ResolveCallback<T> resolve, RejectCallback reject);
+  }
+
+  /**
+   * CancellationException is a helper exception to short-circuit the remainder
+   * of a Promise chain in the event that some precondition fails earlier in the
+   * chain.
+   */
+  @JsType
+  public static class CancellationException extends IllegalStateException {
+    @SuppressWarnings("unused")  // Needed by Serializable
+    CancellationException() {
+      this("");
+    }
+
+    CancellationException(String reason) {
+      super(reason);
+    }
+  }
+
+  /**
+   * WrappedException wraps an arbitrary exception as a RuntimeException so that
+   * it can be passed back through the JSNI layer.
+   */
+  @JsType
+  public static class WrappedException extends RuntimeException {
+    @SuppressWarnings("FieldMayBeFinal")
+    private Throwable original;
+
+    @SuppressWarnings("unused")  // Needed by Serializable
+    WrappedException() {
+      this(null);
+    }
+
+    @JsConstructor
+    public WrappedException(Throwable t) {
+      original = t;
+    }
+
+    /**
+     * Retrieves the underlying exception wrapped by this exception.
+     * @return the underlying exception
+     */
+    public Throwable getOriginal() {
+      return original;
+    }
+  }
+
+  /**
+   * AsyncCallbackProxy provides an interface between the GWT AsyncCallback
+   * interface used for RPC and Promises.
+   *
+   * @param <T> the type of the resolved value
+   */
+  private static class AsyncCallbackProxy<T> extends OdeAsyncCallback<T> {
+    private ResolveCallback<T> resolve;
+    private RejectCallback reject;
+
+    AsyncCallbackProxy(String message) {
+      super(message);
+    }
+
+    @Override
+    public void onFailure(Throwable caught) {
+      // This conditional is required otherwise we will show the stale session
+      // dialog, but in reality the user isn't logged in yet.
+      if (Ode.getInstance().getUser() != null) {
```

Comment:
@ewpatton I'm trying to understand why this can't be handled in `OdeAsyncCallback`. The context is that @jisqyv asked me to work on conversion from gwt rpc to rest.

Reply:
It likely can. This PR only migrated some of the RPCs to use promises, not all. I did not want to make the change in OdeAsyncCallback because that would effect every RPC and I deemed that outside the scope of what I wanted to accomplish at that time.